MaxText/layers/linears.py

```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
     return wrapper(inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel)
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
+    # shape of top_k_weights & top_k_indices: (batch, sequence, exp_per_token)
     top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
-    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
-
-    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
-    softmax_probs = softmax_probs.reshape(gate_logits.shape[:-1] + (self.num_experts_per_tok,))
+    softmax_probs = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
```

Comment:
Instead of weight_dtype shouldn't this be activation dtype?

Reply:
oh I see you are not changing the dtype in this PR. But I think the dtype manipulation here and also here https://github.com/google/maxtext/blob/748305d5bd60d326c98d00b2274515f20b5968d9/MaxText/layers/linears.py#L338 should be changed to cast to float32 before softmax, then back to activation dtype. Perhaps the float32 cast should be configurable